### DISCUSSION TITLE
Building in Windows

### DISCUSSION THREAD

**Discussion by AnthonyPaulO** [cesanta/mongoose#1632] [Ask questions, get answers] [answered]
I've been unable to "make" this under WIndows (installed docket), gives me "Error response from daemon: the working directory c:/some-path-here is invalid, it needs to be an absolute path." Then I had to tweak the paths and make them absolute and then it gave me compile errors.

**Comment by cpq** [comment, marked as answer]
Please follow the https://mongoose.ws/tutorials/tools/

**Comment by abdaziznet** [comment]
Can it run on Windows XP?

**Comment by cpq** [comment]
@abdaziznet 
It can run even on Windows95.

**Comment by gvanem** [comment]
> Please follow the https://mongoose.ws/tutorials/tools/

Using `gcc` to build on Windows is a real PITA. I have Cygwin's `gcc` first in the `PATH` (for other reasons like
`sh`, `sed`, `date` etc. in my GNU-makefiles for MSVC/clang-cl). So a:
```
md tmp
cd tmp
git clone https://github.com/cesanta/mongoose .  (! and yes I get the symlinks created fine)
cd examples\http-server
make
```

fails with:
```
./example
e83948b 1 sock.c:198:mg_open_listener   socket: 0
e83948b 1 net.c:190:mg_listen           Failed: http://0.0.0.0:8000, errno 0
e83948c 1 main.c:88:main                Cannot listen on http://0.0.0.0:8000. Use http://ADDR:PORT or :PORT
```

Arg! Then modifying `Makefile` to:
```diff
--- a/examples/http-server/Makefile 2023-03-01 15:49:00
+++ b/examples/http-server/Makefile 2024-01-22 12:14:20
 ifeq ($(OS),Windows_NT)   # Windows settings. Assume MinGW compiler. To use VC: make CC=cl CFLAGS=/MD OUT=/Feprog.exe
   PROG ?= example.exe           # Use .exe suffix for the binary
-  CC = gcc                      # Use MinGW gcc compiler
+  CC ?= gcc                     # Use MinGW gcc compiler
   CFLAGS += -lws2_32            # Link against Winsock library
   DELETE = cmd /C del /Q /F /S  # Command prompt command to delete files
   OUT ?= -o $(PROG)             # Build output
```

and a `make CC=f:/MinGW32/TDM-gcc/bin/gcc` works almost:
```
gcc.exe: error: %H:%M:%S -DMG_ENABLE_LINES=1: Invalid argument
gnumake: *** [makefile:23: example] Error 1
```

Only with `CFLAGS_MONGOOSE += -DMG_HTTP_DIRLIST_TIME_FMT='"%Y/%m/%d %H:%M:%S"'`, this
TDM-gcc version of `example.exe` works fine.

Lesson I've learned;  stay far away of GCC on Windows.

**Comment by scaprile** [comment]
> stay far away of GCC on Windows

Well, in my very personal opinion, I would remove 'GCC on' from that phrase. :wink:

Now seriously, I would write a script to keep Cygwin from your path. In my case, I use the MinGW icons to start a shell where I do some example development or changes. GCC is fine, and MinGW is great, just keep Cygwin confined to where it is needed (or both).

BTW, we use the 64-bit version of MinGW.
